wcfsetup/install/files/ts/WoltLabSuite/Core/Form/Builder/Field/Controller/Rating.ts

```diff
@@ -0,0 +1,133 @@
+/**
+ * Handles the JavaScript part of the rating form field.
+ *
+ * @author  Matthias Schmidt
+ * @copyright 2001-2020 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @module  WoltLabSuite/Core/Form/Builder/Field/Controller/Rating
+ * @since 5.2
+ */
+
+import * as Core from "../../../../Core";
+import * as Environment from "../../../../Environment";
+
+class Rating {
+  protected readonly _activeCssClasses: string[];
+  protected readonly _defaultCssClasses: string[];
+  protected readonly _field: HTMLElement;
+  protected readonly _input: HTMLInputElement;
+  protected readonly _ratingElements: Map<string, HTMLElement>;
+
+  constructor(fieldId: string, value: string, activeCssClasses: string[], defaultCssClasses: string[]) {
+    this._field = document.getElementById(fieldId + "Container")!;
+    if (this._field === null) {
+      throw new Error("Unknown field with id '" + fieldId + "'");
+    }
+
+    this._input = document.createElement("input");
+    this._input.id = fieldId;
+    this._input.name = fieldId;
+    this._input.type = "hidden";
+    this._input.value = value;
+    this._field.appendChild(this._input);
+
+    this._activeCssClasses = activeCssClasses;
+    this._defaultCssClasses = defaultCssClasses;
+
+    this._ratingElements = new Map();
+
+    const ratingList = this._field.querySelector(".ratingList")!;
+    ratingList.addEventListener("mouseleave", () => this._restoreRating());
+
+    ratingList.querySelectorAll("li").forEach((listItem) => {
+      if (listItem.classList.contains("ratingMetaButton")) {
+        listItem.addEventListener("click", (ev) => this._metaButtonClick(ev));
+        listItem.addEventListener("mouseenter", () => this._restoreRating());
+      } else {
+        this._ratingElements.set(listItem.dataset.rating!, listItem);
+
+        listItem.addEventListener("click", (ev) => this._listItemClick(ev));
+        listItem.addEventListener("mouseenter", (ev) => this._listItemMouseEnter(ev));
+        listItem.addEventListener("mouseleave", () => this._listItemMouseLeave());
+      }
+    });
+  }
+
+  /**
+   * Saves the rating associated with the clicked rating element.
+   */
+  _listItemClick(event: Event): void {
```

Comment:
Should this be private / protected? Applies likewise to the other methods and possibly other modules.